src/hooks/useOnboard.ts

```diff
@@ -0,0 +1,125 @@
+import React from "react";
+import { Wallet } from "bnc-onboard/dist/src/interfaces";
+import Onboard from "bnc-onboard";
+import { Initialization } from "bnc-onboard/dist/src/interfaces";
+
+import { useConnection } from "./useConnection";
+
+const infuraId = process.env.REACT_APP_PUBLIC_INFURA_ID;
+
+const getNetworkName = (chainId: number) => {
+  switch (chainId) {
+    case 1: {
+      return "mainnet";
+    }
+    case 42: {
+      return "kovan";
+    }
+    case 3: {
+      return "ropsten";
+    }
+    case 4: {
+      return "rinkeby";
+    }
+  }
+};
+
+function onboardBaseConfig(_chainId?: number): Initialization {
+  const chainId = _chainId ?? 1;
+  const infuraRpc = `https://${getNetworkName(
+    chainId
+  )}.infura.io/v3/${infuraId}`;
+
+  return {
+    dappId: process.env.REACT_APP_ONBOARD_API_KEY || "",
+    hideBranding: true,
+    networkId: 1, // Default to main net. If on a different network will change with the subscription.
+    walletSelect: {
+      wallets: [
+        { walletName: "metamask", preferred: true },
+        {
+          walletName: "imToken",
+          rpcUrl:
+            chainId === 1
+              ? "https://mainnet-eth.token.im"
+              : "https://eth-testnet.tokenlon.im",
+          preferred: true,
+        },
+        { walletName: "coinbase", preferred: true },
+        {
+          walletName: "portis",
+          apiKey: process.env.NEXT_PUBLIC_PORTIS_API_KEY,
+        },
+        { walletName: "trust", rpcUrl: infuraRpc },
+        { walletName: "dapper" },
+        {
+          walletName: "walletConnect",
+          rpc: { [chainId || 1]: infuraRpc },
+        },
+        { walletName: "gnosis" },
+        { walletName: "walletLink", rpcUrl: infuraRpc },
+        { walletName: "opera" },
+        { walletName: "operaTouch" },
+        { walletName: "torus" },
+        { walletName: "status" },
+        { walletName: "unilogin" },
+        {
+          walletName: "ledger",
+          rpcUrl: infuraRpc,
+        },
+      ],
+    },
+    walletCheck: [
+      { checkName: "connect" },
+      { checkName: "accounts" },
+      { checkName: "network" },
+      { checkName: "balance", minimumBalance: "0" },
+    ],
+    // To prevent providers from requesting block numbers every 4 seconds (see https://github.com/WalletConnect/walletconnect-monorepo/issues/357)
+    blockPollingInterval: 1000 * 60 * 60,
+  };
+}
+
+export function useOnboard() {
+  const { connect, disconnect, update, setError } = useConnection();
+  const instance = React.useMemo(
+    () =>
+      Onboard({
```

Comment:
i am wary of initializing onboard wtihin a react component like this. this was one of the issues with across early on, onboard would be initialized multiple times during the life of the app.  one simple way to get around this is to just bridge these events by initializing onboard outside of react with an event emitter, then subscribing in your context. [heres the example](https://github.com/across-protocol/across-frontend/blob/master/src/utils/onboard.ts) in across, though it uses redux

Heres a really simplified example:

```js
// in onboard.ts
import Events from 'events'
export const events = new Events()
export const onboard = Onboard({
  subscriptions:{
    address:(...args)=>events.emit('address',...args)
    network:(...args)=>events.emit('network',...args)
    ...etc
  } 
})
export const connect => {
  onboard.walletCheck() ...etc
}
export const reset => {
  onboard.reset() ...etc
}

----------------------------
// in your context
import {events,onboard} from './onboard'
Context {
  const [state, dispatch] = React.useReducer(reducer, {});
  useEffect(()=>{
    events.on('address', (...update)=>dispatch({
          type: ActionType.UPDATE_FROM_ERROR,
          payload: { ...update },
        });
   },[])
}

----------------------
// in your navbar
import {connect,reset} from './onboard'

Navbar {
  <ConnectButton onClick={connect}/>
}
```

What do u think, do u think this would be more robust or improved?

Reply:
This implementation is similar to our Umaverse one.

I mean we could shift to this one, I'd have to think about robustness. In the end, you'd need to inject the data into some kind of react thing (Context, reducer). The actual function could be removed from the hook if you prefer.